### PR TITLE
Use the `materialscloud-iframe` template instead

### LINKED ISSUE #12: Content-Security-Policy prevents iframing
I'm not sure which repository is the right place to post this, but for some reason, the optimadeclient tool is sending `Content-Security-Policy: frame-ancestors 'self'; report-uri /optimadeclient/api/security/csp-report` headers:

(note: I'm checking on dev-tools to be sure this has nothing to do with dokku)
```
$ curl -X GET -H "Origin: http://www.materialscloud.org" -I https://dev-tools.materialscloud.org/optimadeclient/
HTTP/1.1 200 OK
Date: Fri, 16 Oct 2020 00:14:31 GMT
Server: TornadoServer/6.0.3
Content-Type: text/html
Content-Security-Policy: frame-ancestors 'self'; report-uri /optimadeclient/api/security/csp-report
Vary: Accept-Encoding
Transfer-Encoding: chunked
```
This prevents iframing on materialscloud.org.

For comparison, these are the headers sent by e.g. seekpath:
```
$ curl -X GET -H "Origin: http://www.materialscloud.org" -I https://dev-tools.materialscloud.org/seekpath/
HTTP/1.1 200 OK
Date: Fri, 16 Oct 2020 00:17:12 GMT
Server: Apache/2.4.29 (Ubuntu)
Content-Length: 14528
Vary: Accept-Encoding
Content-Type: text/html; charset=utf-8
```

### LINKED ISSUE #14: Wrong configuration keyword "cull_idle_timeout"
The configuration keyword in the configuration JSON is wrong, it should be `"cull_idle_timeout"` not `"cull_idle_time"`.

### PR DESCRIPTION
**Pull request by CasperWA**
This also fixes #12 by running the Voilà tornado server from a stand-alone script.
This is to ensure the `--Voila.tornado_settings` option's value is parsed correctly.

**Note**: This currently installs the template by pointing to the [`lite`](https://github.com/materialscloud-org/voila-materialscloud-template/tree/lite) branch of the materialscloud-org/voila-materialscloud-template repository. This should be changed back to pointing to a released version once materialscloud-org/voila-materialscloud-template#16 has been merged and a new version has been released.

This branch has been released on the `dev-dokku` server (http://optimadeclient.matcloud.xyz).

Fixes #14 by updating the keyword accordingly.

### CODE REVIEW

**Comment by CasperWA** [comment]
e26d508 updates the header to not allow _any_ host source, i.e., using the wildcard (`'*'`) as a source, but rather use any subdomain of materialscloud.org, i.e., `https://*.materialscloud.org`.

**Comment by CasperWA** [comment]
@ltalirz I specifically changed it away from the wildcard (`*`) to avoid anyone being able to iframe the tool. You are essentially reverting e26d508 with your commit.

**Comment by ltalirz** [comment]
@CasperWA Exactly, please see my comment in the issue

File: jupyter_config.json [hunk 1]
```diff
@@ -1,13 +1,20 @@
 {
+    "Voila": {
+        "tornado_settings": {
+            "headers": {
```
Comment (ltalirz) [inline jupyter_config.json:4]: does this work or not?

i.e. can these options be removed from the `run.sh` script?

Comment (CasperWA) [reply jupyter_config.json:4]: This does indeed seem to work.

Comment (CasperWA) [reply jupyter_config.json:4]: The issue here being that this JSON file is actually never used when running on dokku, unless it is first copied to the local Jupyter config dir and renamed to `voila.json`.

Comment (CasperWA) [reply jupyter_config.json:4]: I will try and instead use the following option:

```console
--Voila.config_file_paths=<list-item-1>...
    Paths to search for voila.(py|json)
    Default: []
```

Comment (ltalirz) [reply jupyter_config.json]: Great!

**Review by ltalirz — APPROVED** [review]
for the aspects that I understand, this looks good to me.

you might want to increase the `cull_idle_timeout` to something like 15 minutes.
It can even be half a day; it should just not be longer than a week ;-)

P.S. The monitor will ping the app once every 10 minutes, i.e. divide whatever time you put and you'll get the minimum number of jupyter kernels running in the app on average (I believe; would of course be nice if the kernel would simply shut down after the icinga check - happy if you can investigate)

**Comment by ltalirz** [comment]
@CasperWA Can you please test iframing it from `localhost` as well?
This is useful to test the frontend https://github.com/materialscloud-org/frontend-work and, as mentioned before, this didn't work for me with this setting.

**Comment by CasperWA** [comment]
> @CasperWA Can you please test iframing it from `localhost` as well?
> This is useful to test the frontend https://github.com/materialscloud-org/frontend-work and, as mentioned before, this didn't work for me with this stricter setting.

I was going to do this by merging it, pushing it to dev-dokku and checking https://dev-www.materialscloud.org/work/tools/optimadeclient

**Comment by ltalirz** [comment]
> I was going to do this by merging it, pushing it to dev-dokku and checking https://dev-www.materialscloud.org/work/tools/optimadeclient

This only checks iframing from materialscloud.org, not from localhost

**Comment by CasperWA** [comment]
> > I was going to do this by merging it, pushing it to dev-dokku and checking https://dev-www.materialscloud.org/work/tools/optimadeclient
> 
> This only checks iframing from materialscloud.org, not from localhost

I have it working on localhost, however, I believe it has to do with the schema _needing_ to be the same. I.e., it didn't work for anything where I had `https` and it was iframing the notebook running on `http`, but as soon as I iframed it on a `http` host, it was fine.

**Comment by CasperWA** [comment]
I'll leave it with the wildcard for now then, but I want to make it more secure ASAP.

**Comment by ltalirz** [comment]
> I have it working on localhost

Great - if you can run an iframe on localhost that fetches the app from optimadeclient.materialscloud.io, then that's all we need (who knows, perhaps there was a slight modification to the rule or I simply messed up something in my tests).

> however, I believe it has to do with the schema _needing_ to be the same. I.e., it didn't work for anything where I had `https` and it was iframing the notebook running on `http`, but as soon as I iframed it on a `http` host, it was fine.

If server A runs https, then indeed browsers will block iframes from server B if server B does not serve the iframe via https as well (which makes sense, since the communication between your browser and server B would not be encrypted).

**Comment by CasperWA** [comment]
Is there a reason why we don't just use HTTPS for matcloud.xyz?

**Comment by ltalirz** [comment]
> Is there a reason why we don't just use HTTPS for matcloud.xyz as well?

We do https://optimadeclient.matcloud.xyz/ ;-)

**Comment by CasperWA** [comment]
> > Is there a reason why we don't just use HTTPS for matcloud.xyz as well?
> 
> We do https://optimadeclient.matcloud.xyz/ ;-)
> 
> P.S. I didn't have a certificate until last week since we needed to move the domain to a different account.

Do you want to try out the more strict solution then first, and afterwards revert to the "allow all" solution?

**Comment by ltalirz** [comment]
> Do you want to try out the more strict solution then first, and afterwards revert to the "allow all" solution?

If you're saying iframing works from localhost, I'm fine with the more strict solution; just wanted to make sure that we can continue to iframe the tools from the local development server when modifying the frontend

**Pull request merged**